sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala

```diff
@@ -304,46 +308,68 @@ final class DataStreamWriter[T] private[sql](ds: Dataset[T]) {
    * @since 3.1.0
    */
   @throws[TimeoutException]
-  def saveAsTable(tableName: String): StreamingQuery = {
-    this.source = SOURCE_NAME_TABLE
+  def table(tableName: String): StreamingQuery = {
```

Comment:
cc @cloud-fan, I want to raise the discussion about the API name(https://github.com/apache/spark/pull/29767#discussion_r500081260) again in this PR.
I still think maybe the previous name is better as the StreamWriter.table returns a `StreamingQuery` directly. I think the naming for `saveAsTable` in DataFrameWriter because it has the API `save`, but DataStreamWriter doesn't. WDYT?
Also cc @HeartSaVioR and @zsxwing

Reply:
Just a renaming proposal here since the 3.1 has not released yet. If you agree I'll submit another PR to change the name, not mess up with this PR.

Reply:
Did you read all comments there?

The name `table` doesn't provide any meaning of "action". It's more natural to understand `table` as syntax sugar of `format("table")` as we provided for `foreach`.

I'm -1 to make the change. If we want to rename, it should bring back my initial proposal, letting `table` as semantically same as `format("table")` with table name.

Reply:
Probably this could be the another opportunity we can revisit #29767 - I think DataFrameWriter and DataStreamWriter is not same, and it's a bit odd to make DataStreamWriter fit to DataFrameWriter. The former has lots of methods and some methods trigger action. The latter only allowed (before #29767) `start` method to trigger action. #29767 broke this and let two methods trigger action.

Once we revisit this I think we should revisit this as well. Thoughts?

Reply:
`saveAsTable` sounds weird to me since `DataStreamWriter` doesn't have a `save` method like `DataFrameWriter`.

I prefer `DataStreamWriter.table` as it's consistent with `DataStreamReader.table`: `DataStreamReader.table` triggers `load` and `DataStreamWrite.table` triggers `start`.

Reply:
Even in DataStreamReader, `table` works differently with others. There're json, orc, parquet, etc which trigger load directly, but they're just shorthand forms of `format(...).load()`. For the lifecycle of DataStreamReader, the original form of usage is `format(...).options(...).load()`, and `table` becomes a thing which cannot do with `format(...).load()`.

I don't think we should make consistent with `DataStreamReader.table`. I think we should also fix `DataStreamReader.table` as well to make consistent with others. After then, `DataStreamWriter.table` should be special form of `format(...)`, like we did for `DataStreamWriter.foreach`, as I initially proposed in #29767 .
(`saveAsTable` was just an output of compromise, and I regret I shouldn't compromise for the public API.)

Both DataStreamReader and DataStreamWriter, the original form of usage has been `format(...).options(...).load()` and `format(...).options(...).start()` for years and it's not ideal to make branches. That warrants new path of API, like we did for DataFrameWriterV2. If we are strongly insist to not add them to the part of new `format`, let's revert both and discuss the design again in dev@ mailing list.

Reply:
I see DataFrameReader and DataFrameWriter have been already mixed up. I won't insist on `DataStreamReader.table` as `DataFrameReader.table` has been living from 1.4.0, but I still don't agree we should have another trigger entry except `start()` for `DataStreamWriter`. That was a mistake and this is a great chance to fix before we release.

Reply:
So my proposal is reverting #29767 and either 1) add `writeStreamTo` (revisiting #29715) or 2) add `table` as specialized format like `foreach` (initial proposal of #29767). If we need enough discuss to get started, I'm even OK to revert first and move it out of 3.1.0.

Reply:
I prefer something like `table`, because:
```
df.writeStream.table("table_name")
```
seems natural to me (though it is missing a `to` to make total sense. `toTable` also makes sense, but then just `table` is shorter. I wouldn't want to use `saveAsTable`, because it's not saving really, it's starting a stream that's continuously saving/updating/overwriting - and the action depends on the `outputMode`.

Reply:
Again I don't just say about the name. I say about the usage pattern - at least for DataStreamWriter, for years `start()` is the only trigger entry starting the query. Isn't it enough reason to have DataStreamWriterV2 for breaking the pattern? Once we have `df.writeStreamTo("table")` everything is in sync with DataFrameWriterV2 (`df.writeTo("table")`) and no confusion.

If we really insist to have two different trigger entries I'm OK to go with `toTable` but not just simply `table`, because it doesn't say intuitively that the method is a terminal operation. We'll have to rely on return type to see whether it's returning StreamingQuery instance - JVM languages will get blessed by IDE, but I'm not quite sure IDE won't guide for all supported non-JVM languages.

Reply:
I think having 2 branches is OK in `DataStreamWriter`. The reason to have `DataFrameWriterV2` is not that `DataFrameWriter` has multiple branches, but because `SaveMode` is a bad API to describe the table write semantic.

About the naming, I don't have a strong preference. `toTable` looks file.

Reply:
Agree with `toTable`. It seems we can reach an agreement for the naming. I'll submit a PR to rename if If there are no further objections. cc @viirya @zsxwing

Reply:
I think it is fine to have another trigger entry. `toTable` looks okay to me. We can describe clearly in the method doc that it will create the table if it doesn't exxist.

Reply:
Thanks everyone for joining the discussion! I created #30571 for the renaming. Will resolve the conversation here.